include/zephyr/drivers/rtc.h

```diff
@@ -0,0 +1,527 @@
+ * @return -EINVAL if id is out of range
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_disable_callback(const struct device *dev, uint16_t id);
+
+static inline int z_impl_rtc_alarm_disable_callback(const struct device *dev, uint16_t id)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_disable_callback(dev, id);
+}
+
+/**
+ * @brief API for setting update callback
+ *
+ * @details The update occurs every time the RTC clock is updated by
+ * 1 second
+ *
+ * @param dev Device instance
+ * @param callback Callback called when update occurs
+ * @param user_data Optional user data passed to callback
+ *
+ * @return 0 if successful
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_update_set_callback(const struct device *dev, rtc_update_callback callback,
+				      void *user_data);
+
+static inline int z_impl_rtc_update_set_callback(const struct device *dev,
+						 rtc_update_callback callback, void *user_data)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->update_set_callback(dev, callback, user_data);
+}
+
+/**
+ * @brief API for enabling RTC update callback
+ *
+ * @param dev Device instance
+ *
+ * @return 0 if successful
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_update_enable_callback(const struct device *dev);
+
+static inline int z_impl_rtc_update_enable_callback(const struct device *dev)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->update_enable_callback(dev);
+}
+
+/**
+ * @brief API for disabling RTC update callback
+ *
+ * @param dev Device instance
+ *
+ * @return 0 if successful
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_update_disable_callback(const struct device *dev);
+
+static inline int z_impl_rtc_update_disable_callback(const struct device *dev)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->update_disable_callback(dev);
+}
+
+/**
+ * @brief API for setting RTC calibration
+ *
+ * @details Calibration is applied to the RTC clock input
+ *
+ * @param dev Device instance
+ * @param calibration Calibration to set in parts per billion
+ *
+ * @return 0 if successful
+ * @return -EINVAL if calibration is out of range
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_set_calibration(const struct device *dev, int32_t calibration);
+
+static inline int z_impl_rtc_set_calibration(const struct device *dev, int32_t calibration)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->set_calibration(dev, calibration);
+}
```

Comment:
Does an RTC need to support this method?  I suspect many won't.

Existing Zephyr APIs will return -ENOSYS if the api-> pointer for the method they use is not defined.  This prevents every driver from needing to make a set of stub methods for the portions they don't impliment.

Suggest do that here, and for any other portion of the API that could be considered optional.

Reply:
I will go as far as stating that the majority RTCs support clock calibration. The RV3032 doesn't, but it is quite an expensive feature rich piece of hardware, costing the same as some SOCs, not exactly the norm for embedded RTCs :)

Reply:
> I will go as far as stating that the majority RTCs support clock calibration

Have used microchip and dallas RTCs for years, never seen one with a calibration feature.

Reply:
The RV3032 does support calibration.  The Swiss not supporting a precise timepiece?  Register 0xc1 contains an EEPROM byte that supports adjustment in 238.4 ppb steps over a range of ±7.4 ppm.  The temperature compensation can also be calibrated via registers 0xc4/0xc5.  I didn't add it in my driver as there was no API for it.

But I often see drivers that don't support it.  The two existing Zephyr RTC drivers, MCP7940n and DS3231 don't support it.  The DS3231 driver has the `aging` register the register map, but doesn't have means to set it.

I think it's worthwhile to consider making part of the API optional for drivers if a significant fraction won't implement it.